File: output/CodeGen/codenet/C++/Python/s338629964.py
#!/usr/bin/python

import sys

def init():
    global fact
    global revfact
    global rev
    fact = [1]
    revfact = [1]
    rev = [0]
    for i in range(1, 1100000):
        fact.append(fact[i-1] * i % 1000000007)
        if i > 1:
            rev.append(1000000007 / i * (1000000007 - rev[i % 1000000007]) % 1000000007)
        revfact.append(revfact[i-1] * rev[i] % 1000000007)

def nCk(n, k):
    if n < k:
        return 0
    if comb[n][k]!= 0:
        return comb[n][k]
    return comb[n][k] = fact[n] * revfact[n-k] % 1000000007 * revfact[k] % 1000000007

def mod_pow(x, a, m = 1000000007):
    if a == 0:
        return 1
    if a % 2 == 0:
        return mod_pow(x, a / 2, m) * mod_pow(x, a / 2, m) % m
    return mod_pow(x, a / 2, m) * mod_pow(x, a / 2, m) * x % m

def inv(x, m = 1000000007):
    return mod_pow(x, m-2, m)

def aall, ball:
    global fact
    global revfact
    global rev
    aall = 0
    ball = 0
    for i in range(1, len(sys.argv)):
        aall += int(sys.argv[i])
        ball += int(sys.argv[i+1])
    return aall, ball

def dp(n, asum, cnt, p, p1):
    global fact
    global revfact
    global rev
    global comb
    if asum < 0 or cnt < 0 or p < 0 or p1 < 0:
        return 0
    if asum == 0 and cnt == 0 and p == 0 and p1 == 0:
        return 1
    if comb[n][cnt]!= 0:
        return comb[n][cnt]
    return comb[n][cnt] = dp(n-1, asum, cnt, p, p1) + dp(n-1, asum-a[n-1], cnt, p^1, p1) * nCk(cnt, b[n-1]-1) * rev[asum] * revfact[asum] * inv(a[n-1]) % 1000000007 * inv(a[n-1]) % 1000000007

def main():
    init()
    n, asum, cnt, p, p1 = map(int, sys.argv[1:])
    ans = dp(n, asum, cnt, p, p1)
    if ans < 0:
        ans += 1000000007
    print ans

if __name__ == '__main__':
    main()

File: output/CodeGen/codenet/C++/Java/s338629964.java
public class s338629964 {
    public static void main(String[] args) {
        int n = sc.nextInt();
        int[] a = new int[n];
        int[] b = new int[n];
        int aall = 0;
        int ball = 0;
        for (int i = 0; i < n; i++) {
            a[i] = sc.nextInt();
            b[i] = sc.nextInt();
            aall += a[i];
            ball += b[i];
        }
        int[][] dp = new int[aall + 1][ball + 1];
        dp[0][0] = 1;
        for (int i = 1; i <= n; i++) {
            int id = i % 2;
            for (int asum = aall; asum >= 0; asum--) {
                for (int cnt = 0; cnt <= ball; cnt++) {
                    for (int p = 0; p < 2; p++) {
                        if (asum - a[i - 1] < 0) continue;
                        int pw = 1;
                        for (int use = 0; use < b[i - 1]; use++) {
                            if (cnt - use >= 0) {
                                int tmp = nCk(cnt, use) * pw * dp[asum - a[i - 1]][cnt - use] % MOD * inv(a[i - 1]) % MOD;
                                dp[asum][cnt][p] += tmp;
                                dp[asum][cnt][p] %= MOD;
                            } else break;
                            pw = pw * a[i - 1] % MOD;
                        }
                        if (cnt - b[i - 1] + 1 >= 0) {
                            dp[asum][cnt][p] += nCk(cnt, b[i - 1] - 1) * pw * dp[asum - a[i - 1]][cnt - b[i - 1] + 1] % MOD * inv(a[i - 1]) % MOD;
                            dp[asum][cnt][p] %= MOD;
                        }
                    }
                }
            }
        }
        int ans = 0;
        for (int asum = 0; asum <= aall; asum++) {
            int rev_asum_pow = 1;
            for (int cnt = 0; cnt <= ball; cnt++) {
                rev_asum_pow = rev_asum_pow * inv(rev[asum]) % MOD;
                for (int p = 0; p < 2; p++) {
                    int x = dp[asum][cnt][p] * rev_asum_pow * (cnt + 1) % MOD * aall % MOD * rev[asum] % MOD;
                    if (p == 1) ans += x;
                    else ans -= x;
                    ans %= MOD;
                }
            }
        }
        if (ans < 0) ans += MOD;
        System.out.println(ans);
    }
}